File: adventofcode/2022/day12/day12.hs
import Data.List (find)
import Data.Char (ord)
import qualified Data.Map as M

data Point = Point Int Int
  deriving (Show, Eq, Ord)

data Grid = Grid 
  { start :: Point
  , end :: Point
  , points :: [String]
  , start_points :: [Point]
  }
  deriving (Show)

parseGrid :: [String] -> Grid
parseGrid ls = Grid start end ls start_points
  where ysize = length ls
        xsize = length $ head ls
        ps = [Point x y | x <- [0..xsize-1], y <- [0..ysize-1]]
        f :: Point -> Char
        f (Point x y) = (ls !! y) !! x
        (Just start) = find ((== 'S') . f) ps
        (Just end) = find ((== 'E') . f) ps
        start_points = start : filter ((== 'a') . f) ps

bounds :: Grid -> (Int, Int)
bounds (Grid { points }) = (xsize, ysize)
  where ysize = length points
        xsize = length $ head points

get :: Grid -> Point -> Int
get (Grid { start, end, points }) p@(Point x y)
  | p == start = ord 'a'
  | p == end = ord 'z'
  | otherwise = ord $ (points !! y) !! x

inside :: Grid -> Point -> Bool
inside grid (Point x y) = x >= 0 && x < xsize && y >= 0 && y < ysize
  where (xsize, ysize) = bounds grid

neighbours :: Grid -> Point -> [Point]
neighbours grid (Point x y) = filter (inside grid) ps
  where ps = [ Point x (y -1)
             , Point (x - 1) y
             , Point (x + 1) y
             , Point x (y + 1)
             ]

dist :: Grid -> Point -> Point -> Int
dist grid p1 p2 = (get grid p2) - (get grid p1)

visitMany :: [Point] -> M.Map Point Bool -> M.Map Point Bool
visitMany [] vis = vis
visitMany (p:ps) vis = visitMany ps (M.insert p True vis)

findShortestPathLength :: Grid -> [(Point, Int)] -> M.Map Point Bool -> Int
findShortestPathLength grid [] vis = maxBound :: Int
findShortestPathLength grid ((p, steps):ps) vis
  | p == end grid = steps
  | otherwise = findShortestPathLength grid (ps ++ ns) nvis
  where ns = [(n, steps + 1) | n <- neighbours grid p, M.notMember n vis, (<= 1) $ dist grid p n]
        nvis = visitMany (map fst ns) vis

part1 :: Grid -> Int
part1 grid = findShortestPathLength grid queue vis
  where vis = M.singleton (start grid) True
        queue = [(start grid, 0)]

part2 :: Grid -> Int
part2 grid = minimum $ map (\s -> findShortestPathLength grid [(s, 0)] (M.singleton s True)) $ start_points grid

main = interact $ unlines . map show . sequence [part1, part2] . parseGrid . lines

File: adventofcode/2022/day12/day12.rb
# frozen_string_literal: true

class Grid
  attr_accessor :grid, :start, :end

  def initialize(lines)
    @grid = lines.split.map(&:chars)

    (0...@grid.size).each do |y|
      (0...@grid[0].size).each do |x|
        @start = [x, y] if @grid[y][x] == 'S'
        @end = [x, y] if @grid[y][x] == 'E'
      end
    end
  end

  def get(point)
    return 'a' if point == @start
    return 'z' if point == @end

    x, y = point
    @grid[y][x]
  end

  def find_char(char)
    ps = []
    (0...@grid.size).each do |y|
      (0...@grid[0].size).each do |x|
        p = [x, y]
        ps.push(p) if self.get(p) == char
      end
    end
    ps
  end

  def inside?(point)
    x, y = point
    xsize = @grid[0].size
    ysize = @grid.size

    x >= 0 && x < xsize && y >= 0 && y < ysize
  end

  def dist(p1, p2)
    x1, y1 = p1
    x2, y2 = p2
    ord1 = p1 == self.start ? 'a'.ord : @grid[y1][x1].ord
    ord2 = p2 == self.end ? 'z'.ord : @grid[y2][x2].ord
    ord2 - ord1
  end

  def neighbours(point)
    x, y = point
    [
      [x, y - 1],
      [x - 1, y],
      [x + 1, y],
      [x, y + 1],
    ].filter {|p| self.inside?(p)}
  end

  def find_shortest_path(pstart, pend)
    q = [[pstart, 0]]
    vis = {}
    vis[pstart] = true
    while !q.empty?
      p, steps = q.shift
      return steps if p == pend

      self.neighbours(p).each do |n|
        if !vis[n] && self.dist(p, n) <= 1
          vis[n] = true
          q.push([n, steps + 1])
        end
      end
    end
    return (2**(0.size * 8 -2) -1)
  end
end

def read_input
  ARGF.read
end

def part1(grid)
  grid.find_shortest_path(grid.start, grid.end)
end

def part2(grid)
  grid.find_char('a').map do |s|
    grid.find_shortest_path(s, grid.end)
  end.min
end

grid = Grid.new(read_input)
p part1 grid
p part2 grid
